sql/core/src/test/scala/org/apache/spark/sql/CollationSuite.scala

```diff
@@ -639,4 +639,169 @@ class CollationSuite extends DatasourceV2SQLBase with AdaptiveSparkPlanHelper {
+          "array('AAA' collate utf8_binary_lcase, 'BBB' collate utf8_binary_lcase)"
+        ),
+        Seq((Seq("aaa", "bbb"), 2))),
+      ("utf8_binary_lcase",
+        Seq(
+          "array('aaa' collate utf8_binary_lcase)",
+          "array('bbb' collate utf8_binary_lcase)",
+          "array('AAA' collate utf8_binary_lcase)",
+          "array('BBB' collate utf8_binary_lcase)"
+        ),
+        Seq((Seq("aaa"), 2), (Seq("bbb"), 2)))
+    )
+    val nested = Seq(
+      // binary
+      ("utf8_binary", Seq("array(array('aaa'))", "array(array('AAA'))"),
+        Seq((Seq(Seq("aaa")), 1), (Seq(Seq("AAA")), 1))),
+      ("utf8_binary", Seq("array(array('aaa'), array('bbb'))", "array(array('AAA'), array('bbb'))"),
+        Seq((Seq(Seq("aaa"), Seq("bbb")), 1), (Seq(Seq("AAA"), Seq("bbb")), 1))),
+      // non-binary
+      ("utf8_binary_lcase",
+        Seq(
+          "array(array('aaa' collate utf8_binary_lcase))",
+          "array(array('AAA' collate utf8_binary_lcase))"
+        ),
+        Seq((Seq(Seq("aaa")), 2))),
+      ("utf8_binary_lcase",
+        Seq(
+          "array(array('aaa' collate utf8_binary_lcase), array('bbb' collate utf8_binary_lcase))",
+          "array(array('AAA' collate utf8_binary_lcase), array('bbb' collate utf8_binary_lcase))"
+        ),
+        Seq((Seq(Seq("aaa"), Seq("bbb")), 2))),
+      ("utf8_binary_lcase",
+        Seq(
+          "array(array('aaa' collate utf8_binary_lcase, 'AAA' collate utf8_binary_lcase)," +
+            "array('bbb' collate utf8_binary_lcase, 'ccc' collate utf8_binary_lcase))",
+          "array(array('aaa' collate utf8_binary_lcase, 'aaa' collate utf8_binary_lcase)," +
+            "array('bbb' collate utf8_binary_lcase, 'ccc' collate utf8_binary_lcase)," +
+            "array('ddd' collate utf8_binary_lcase))",
+          "array(array('AAA' collate utf8_binary_lcase, 'aaa' collate utf8_binary_lcase)," +
+            "array('BBB' collate utf8_binary_lcase, 'CCC' collate utf8_binary_lcase))"
+        ),
+        Seq(
+          (Seq(Seq("aaa", "AAA"), Seq("bbb", "ccc")), 2),
+          (Seq(Seq("aaa", "aaa"), Seq("bbb", "ccc"), Seq("ddd")), 1)
+        )
+      )
+    )
+
+    val all = simple.map {
+      case (collName, data, res) => (s"array<string collate $collName>", data, res)
+    } ++ nested.map {
+      case (collName, data, res) => (s"array<array<string collate $collName>>", data, res)
+    }
+
+    all.foreach {
+      case (dt, rows, count) =>
+        withTable(tableName) {
+          sql(s"create table $tableName(a $dt) using parquet")
+          rows.map(row => sql(s"insert into $tableName(a) values($row)"))
+          checkAnswer(sql(s"select a, count(*) from $tableName group by a"),
+            count.map { case (aggStr, cnt) => Row(aggStr, cnt) })
+          checkAnswer(sql(s"select distinct a from $tableName"),
+            count.map{ case (aggStr, _) => Row(aggStr)})
+        }
+    }
+  }
```

Comment:
The test pattern that can be useful here is what we did for Window Aggs.
In short, just create a query that targets mixed-case data with LCASE collation (e.g. "aA", "aa", "AA") and query that targets normalized data with UTF8_BINARY ("aa", "aa", "aa"). Aggs and Joins should return the same result.

You can find test example here:
https://github.com/apache/spark/pull/45568

Reply:
Updated tests to support this and added case classes for better readability. Please check again.